xdsl/transforms/canonicalization_patterns/arith.py

```diff
@@ -16,3 +16,48 @@ def match_and_rewrite(self, op: arith.Addi, rewriter: PatternRewriter) -> None:
             and value.value.data == 0
         ):
             rewriter.replace_matched_op([], [op.rhs])
+
+
+class SelectConstPattern(RewritePattern):
+    """
+    arith.select %true %x %y = %x
+    arith.select %false %x %y = %y
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: arith.Select, rewriter: PatternRewriter):
+        if not isinstance(condition := op.cond.owner, arith.Constant):
+            return
+
+        assert isinstance(const_cond := condition.value, IntegerAttr)
+
+        if const_cond.value.data == 1:
+            rewriter.replace_matched_op((), (op.lhs,))
+        if const_cond.value.data == 0:
+            rewriter.replace_matched_op((), (op.rhs,))
+
+
+class SelectTrueFalsePattern(RewritePattern):
+    """
+    arith.select %x %true %false = %x
+    arith.select %x %false %true = %x xor 1
```

Comment:
For an `i1` result type you may even consider folding the `%true %true` and `%false %false` cases if you think they may realistically come up.

Reply:
More generally, `arith.select %x %y %y` should be %y right? I realise I didn't include this as I didn't immediately find it in mlir but it would probably be nice to add